histogram.go

```diff
@@ -0,0 +1,254 @@
+	return []time.Duration(v)
+}
+
+// BucketPairs creates a set of bucket pairs from a set
+// of buckets describing the lower and upper bounds for
+// each derived bucket.
+func BucketPairs(buckets Buckets) []BucketPair {
+	var (
+		asValueBuckets    = buckets.AsValues()
+		asDurationBuckets = buckets.AsDurations()
+	)
+	var pairs []BucketPair
+	if buckets.Len() > 0 {
+		pairs = append(pairs, bucketPair{
+			-math.MaxFloat64, asValueBuckets[0],
+			time.Duration(math.MinInt64), asDurationBuckets[0]})
+
+		prevValueBucket, prevDurationBucket :=
+			asValueBuckets[0], asDurationBuckets[0]
+		for i := 1; i < buckets.Len(); i++ {
+			pairs = append(pairs, bucketPair{
+				prevValueBucket, asValueBuckets[i],
+				prevDurationBucket, asDurationBuckets[i]})
+			prevValueBucket, prevDurationBucket =
+				asValueBuckets[i], asDurationBuckets[i]
+		}
+
+		pairs = append(pairs, bucketPair{
+			prevValueBucket, math.MaxFloat64,
+			prevDurationBucket, time.Duration(math.MaxInt64)})
+	} else {
+		pairs = append(pairs, bucketPair{
+			-math.MaxFloat64, math.MaxFloat64,
+			time.Duration(math.MinInt64), time.Duration(math.MaxInt64)})
+	}
+	return pairs
+}
+
+type bucketPair struct {
+	lowerBoundValue    float64
+	upperBoundValue    float64
+	lowerBoundDuration time.Duration
+	upperBoundDuration time.Duration
+}
+
+func (p bucketPair) LowerBoundValue() float64 {
+	return p.lowerBoundValue
+}
+
+func (p bucketPair) UpperBoundValue() float64 {
+	return p.upperBoundValue
+}
+
+func (p bucketPair) LowerBoundDuration() time.Duration {
+	return p.lowerBoundDuration
+}
+
+func (p bucketPair) UpperBoundDuration() time.Duration {
+	return p.upperBoundDuration
+}
+
+// LinearValueBuckets creates a set of linear value buckets.
+func LinearValueBuckets(start, width float64, count int) Values {
+	if count <= 0 {
+		panic("count needs to be > 0")
+	}
+	buckets := make([]float64, count)
+	for i := range buckets {
+		buckets[i] = start + (float64(i) * width)
+	}
+	return Values(buckets)
+}
+
+// LinearDurationBuckets creates a set of linear duration buckets.
+func LinearDurationBuckets(start, width time.Duration, count int) Durations {
+	if count <= 0 {
+		panic("count needs to be > 0")
+	}
+	buckets := make([]time.Duration, count)
+	for i := range buckets {
+		buckets[i] = start + (time.Duration(i) * width)
+	}
+	return Durations(buckets)
+}
+
+// ExponentialValueBuckets creates a set of exponential value buckets.
+func ExponentialValueBuckets(start, factor float64, count int) Values {
+	if count <= 0 {
+		panic("count needs to be > 0")
+	}
+	if start <= 0 {
+		panic("start needs to be > 0")
+	}
+	if factor <= 1 {
+		panic("factor needs to be > 1")
+	}
+	buckets := make([]float64, count)
+	curr := start
+	for i := range buckets {
+		buckets[i] = curr
+		curr *= factor
+	}
+	return Values(buckets)
+}
+
+// ExponentialDurationBuckets creates a set of exponential duration buckets.
+func ExponentialDurationBuckets(start time.Duration, factor float64, count int) Durations {
```

Comment:
So it won't actually allocate each time you call `scope.Histogram`.  The interface type `tally.Buckets` is allocated on the stack in all locations during histogram creation.  The histogram field `specification tally.Buckets` does not take a reference to the stack allocated interface passed in, sets the interface type with the `tab` and `data fields by value:
https://golang.org/src/runtime/runtime2.go?h=iface#L143

Take the following example:
```go
package main

import "testing"

var existingFooBar = &fooBar{}

type Foo interface {
	Foo()
}

type FooBar interface {
	Foo()
	Bar()
}

type fooBar struct {
	x int
}

func (*fooBar) Foo() {}
func (*fooBar) Bar() {}

type thing struct {
	foo Foo
}

func newThing(foo Foo) thing {
	return thing{foo: foo}
}

func BenchmarkIfaceAllocs(b *testing.B) {
	ref := existingFooBar
	for i := 0; i < b.N; i++ {
		x := newThing(ref)
		if x.foo != ref {
			b.Fail()
		}
//		ref = &fooBar{x: i}
	}
}
```

When run with `ref = &fooBar{x: i}` _not_ commented out it causes one allocation per op:
```
$ go test -bench Iface -v -benchmem
BenchmarkIfaceAllocs-4          100000000               17.6 ns/op             8 B/op            1 allocs/op
PASS
```

When run with `ref = &fooBar{x: i}` commented out it cases zero allocations per op:
```
$ go test -bench Iface -v -benchmem
BenchmarkIfaceAllocs-4          2000000000               1.28 ns/op            0 B/op            0 allocs/op
PASS
```